Added a new section for FUNCTION apps, including index, recommendations and KQL files

parthshiras opened the pull request:
<!-- Thank you for submitting a Pull Request. Please fill out the template below.-->

# Overview/Summary

Added a new section for function apps, including index, recommendations and KQL files

## Related Issues/Work Items

1. [AB#29861](https://dev.azure.com/CSUSolEng/12ba7e46-a92a-47f6-ad6a-fba0479234a7/_workitems/edit/29861)
2. [AB#29857](https://dev.azure.com/CSUSolEng/12ba7e46-a92a-47f6-ad6a-fba0479234a7/_workitems/edit/29857)
3. [AB#29860](https://dev.azure.com/CSUSolEng/12ba7e46-a92a-47f6-ad6a-fba0479234a7/_workitems/edit/29860)
4. [AB#29858](https://dev.azure.com/CSUSolEng/12ba7e46-a92a-47f6-ad6a-fba0479234a7/_workitems/edit/29858)
5. [AB#29859](https://dev.azure.com/CSUSolEng/12ba7e46-a92a-47f6-ad6a-fba0479234a7/_workitems/edit/29859)

## This PR fixes/adds/changes/removes

1. azure-resources/Web/functions/_index.md -> Added for new subsection for Functions within Web
2. azure-resources/Web/functions/kql/0b06a688-0dd6-4d73-9f72-6666ff853ca9.kql -> Added KQL file
3. azure-resources/Web/functions/kql/52f368ee-1d77-4b34-92db-64be269642d0.kql -> Added KQL file
4. azure-resources/Web/functions/kql/7c608f46-46b2-4cc0-bbd6-1d457c16671c.kql -> Added KQL file
5. azure-resources/Web/functions/kql/c9a278b7-024b-454b-bd54-41587c512b74.kql -> Added KQL file
6. azure-resources/Web/functions/kql/c6c4b962-5af4-447a-9d74-7b9c53a5dff5.kql -> Added KQL file for checking Auto Heal settings
7. azure-resources/Web/functions/recommendations.yaml -> Added markup file

### Breaking Changes

None

## As part of this pull request I have

- [x] Read the [Contribution Guide](https://azure.github.io/Azure-Proactive-Resiliency-Library-v2/contributing) and ensured this PR is compliant with the guide
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Azure-Proactive-Resiliency-Library-v2/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/Azure-Proactive-Resiliency-Library-v2/issues) or ADO Work Items (Internal Only)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Azure-Proactive-Resiliency-Library-v2/tree/main)
- [ ] Ensured PR tests are passing
- [x] Performed testing and provided evidence (e.g. screenshot of output) for any changes associated to ARG queries
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Docs etc.)

![image](https://github.com/Azure/Azure-Proactive-Resiliency-Library-v2/assets/18480893/b8d62bb3-eb7e-44a9-857c-2b5196c8a4ec)

ejhenry reviewed (commented):
Thanks for the contribution @parthshiras. Please ensure the recommendation description is worded as a recommendation, the longDescription field is within the character limit, and only a single learn more link is provided.

parthshiras commented:
> Thanks for the contribution @parthshiras. Please ensure the recommendation description is worded as a recommendation, the longDescription field is within the character limit, and only a single learn more link is provided.



> Thanks for the contribution @parthshiras. Please ensure the recommendation description is worded as a recommendation, the longDescription field is within the character limit, and only a single learn more link is provided.



> Thanks for the contribution @parthshiras. Please ensure the recommendation description is worded as a recommendation, the longDescription field is within the character limit, and only a single learn more link is provided.

Reviewed all long descriptions and updated to be within the 300 character limit.

parthshiras commented:
> @parthshiras do you have questions about the feedback or require assistance with the changes?



> @parthshiras do you have questions about the feedback or require assistance with the changes?



> @parthshiras do you have questions about the feedback or require assistance with the changes?

Resolved open items. Thanks for the review Eric!

ejhenry commented on azure-resources/Web/functions/kql/0b06a688-0dd6-4d73-9f72-6666ff853ca9.kql:
```diff
@@ -0,0 +1 @@
+// cannot-be-validated-with-arg
```
```suggestion
// cannot-be-validated-with-arg

```

ejhenry commented on azure-resources/Web/functions/kql/52f368ee-1d77-4b34-92db-64be269642d0.kql:
```diff
@@ -0,0 +1 @@
+// cannot-be-validated-with-arg
```
```suggestion
// cannot-be-validated-with-arg

```

ejhenry commented on azure-resources/Web/functions/kql/7c608f46-46b2-4cc0-bbd6-1d457c16671c.kql:
```diff
@@ -0,0 +1 @@
+// cannot-be-validated-with-arg
```
```suggestion
// cannot-be-validated-with-arg

```

ejhenry commented on azure-resources/Web/functions/kql/c6c4b962-5af4-447a-9d74-7b9c53a5dff5.kql:
```diff
@@ -0,0 +1,13 @@
+    | where properties.AutoHealEnabled == 'false'
+    | project id, name, tenantId, location, resourceGroup, properties.AutoHealEnabled
+    ) on name
+| project recommendationID = "c6c4b962-5af4-447a-9d74-7b9c53a5dff5", name, id, type, kind, param1="AutoHealEnabled =false"
```
```suggestion
| project recommendationID = "c6c4b962-5af4-447a-9d74-7b9c53a5dff5", name, id, type, kind, param1="AutoHealEnabled =false"

```

ejhenry commented on azure-resources/Web/functions/kql/c9a278b7-024b-454b-bd54-41587c512b74.kql:
```diff
@@ -0,0 +1 @@
+// cannot-be-validated-with-arg
```
```suggestion
// cannot-be-validated-with-arg

```

ejhenry reviewed (approved):
LGTM

oZakari reviewed (changes requested):
@ejhenry @parthshiras  As the resource type for Azure Functions falls under Microsoft.Web/sites, we should move these recommendations to existing recommendations file under Web/sites. We can specify in the description that these align specifically to functions.

aarthiem commented:
@ejhenry - I would recommend a separate folder for functions be created under sites and place the recommendations. Similarly for other services like Logic Apps.

ejhenry commented:
> @ejhenry - I would recommend a separate folder for functions be created under sites and place the recommendations. Similarly for other services like Logic Apps.

@aarthiem thanks for the feedback. We need to align with our current APRL standard of grouping recommendations by resource provider and top-level resource type that @oZakari outlined in his comment. We rely on this standard for our automation. Do you think that there will still be confusion if the recommendation description clearly states it is applicable to Functions?

ejhenry commented:
@parthshiras let me know if you have any questions about the change requested by @oZakari

parthshiras commented:
> @parthshiras let me know if you have any questions about the change requested by @oZakari

Hi Eric - yes connected with Aarthi, and will make changes per suggestions.

parthshiras commented:
@ejhenry - hi Eric, I have moved all functions related files and recommendations to web/sites per the recommendation and deleted the directory.

cc - @aarthiem

ejhenry reviewed (approved):
Thanks @parthshiras LGTM 👍

oZakari reviewed (approved):
Great work @parthshiras, thank you for your contribution!
